### PR TITLE
[Terrain] Unit tests for terrain world component settings

### PR DESCRIPTION
**Pull request by mbalfour-amzn**
## What does this PR do?

This adds unit tests for the Terrain World Component and the settings that it controls. The unit tests for the Terrain World Component itself just verify that they set the settings correctly on the TerrainSystem. There are now separate unit tests for the TerrainSystem to test that the settings themselves are functional - World Min/Max, Height Query Resolution, and Surface Data Query Resolution.

While working on the unit tests, I discovered a number of problems that needed to be addressed as well:
* Surface Data Query Resolution and the sampler type were both being ignored by the Surface Weight queries, which has been fixed. The CLAMP and BILINEAR types both now clamp the position being queried to the Surface Query Resolution.
* Added SurfaceTag ToString method so that it can print successfully from python logging
* Removed offset from TerrainDetailManager, it was offsetting the data incorrectly, causing the detail materials not to line up with the macro material
* Changed CLAMP height sampler type to round the position to nearest, instead of always rounding down.
* Changed BILINEAR height sampler type to better handle the cases where one or more of the 4 points didn't exist.
* Fixed support for the "Use Ground Plane" checkbox on Terrain Layer Spawners
* Fixed some cases where the singular query methods produced different results than the multi-query methods
* Changed the Terrain Layer Spawners to use an AABB that's min-inclusive-max-exclusive so that edges shared between two spawners have one clear owner. Previously they included the max edge as well, making edge overlaps have an owner based either on priority or non-deterministic luck, in the case of equal priority.
* The "TerrainSystem_VegetationSpawnsOnTerrainSurfaces" was using fragile logic that's been made more robust (and commented) to handle all of the bugfixes listed above.
* All of the existing terrain unit tests have been updated to match the changes listed above.

## How was this PR tested?

Ran all the existing unit tests and verified that they still work. Verified that the new unit tests work. Verified that the new unit tests that test Surface Data Query Resolution were failing until the Terrain System was fixed to use the sampler type. Verified that the terrain automated tests still work.

### CODE REVIEW

**Review by cgalvan — APPROVED** [review]

**Review by invertednormal — COMMENTED** [review]
Looks good!

**Review by invertednormal — APPROVED** [review]

File: Gems/SurfaceData/Code/Include/SurfaceData/Utility/SurfaceDataUtility.h [hunk 1]
```diff
@@ -144,4 +144,23 @@ namespace SurfaceData
...
+    }
+    // Utility method to compare an AABB and a point for overlapping XY coordinates while ignoring the Z coordinates.
+    // This method includes points that land on the min edge but excludes points that land on the max edge.
+    AZ_INLINE bool AabbContains2DMaxExclusive(const AZ::Aabb& box, const AZ::Vector3& point)
+    {
+        return box.GetMin().GetX() <= point.GetX() &&
+               box.GetMin().GetY() <= point.GetY() &&
+               box.GetMax().GetX() > point.GetX() &&
+               box.GetMax().GetY() > point.GetY();
+    }
```
Comment (invertednormal) [inline Gems/SurfaceData/Code/Include/SurfaceData/Utility/SurfaceDataUtility.h:165]: These are basically the same function. I'd have the second one call
```AabbContains2DMaxExclusive(box, AZ::Vector2(point))```
Another option is to just template on the vector type.

**Review by invertednormal — APPROVED** [review]
Looks good, just one suggestion to reduce code duplication.

File: Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp [hunk 2]
```diff
@@ -223,6 +223,72 @@ void TerrainSystem::RoundPosition(float x, float y, float queryResolution, AZ::V
     outPosition = (normalizedPosition + AZ::Vector2(0.5f)).GetFloor() * queryResolution;
 }
 
+void TerrainSystem::InterpolateHeights(float heights[4], bool exists[4], float lerpX, float lerpY, float& outHeight, bool& outExists)
```
Comment (invertednormal) [inline Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:226]: `heights` and `exists` could be `const`. Would be a little cleaner to use const refs to `AZStd::array<>`.

File: Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp [hunk 3]
```diff
@@ -480,13 +562,15 @@ float TerrainSystem::GetHeightSynchronous(float x, float y, Sampler sampler, boo
...
-            const float heightXY0 = AZ::Lerp(heightX0Y0, heightX1Y0, normalizedDelta.GetX());
-            const float heightXY1 = AZ::Lerp(heightX0Y1, heightX1Y1, normalizedDelta.GetX());
-            height = AZ::Lerp(heightXY0, heightXY1, normalizedDelta.GetY());
+            bool exists[4] = { false, false, false, false };
+            float queriedHeights[4] = { 0.0f, 0.0f, 0.0f, 0.0f };
+
+            queriedHeights[0] = GetTerrainAreaHeight(pos0.GetX(), pos0.GetY(), exists[0]);
+            queriedHeights[1] = GetTerrainAreaHeight(pos1.GetX(), pos0.GetY(), exists[1]);
+            queriedHeights[2] = GetTerrainAreaHeight(pos0.GetX(), pos1.GetY(), exists[2]);
+            queriedHeights[3] = GetTerrainAreaHeight(pos1.GetX(), pos1.GetY(), exists[3]);
```
Comment (invertednormal) [inline Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp:571]: I'd just combine these so queriedHeights could be const - 
```
const float queriedHeights[4] = {
    GetTerrainAreaHeight(pos0.GetX(), pos0.GetY(), exists[0]),
    GetTerrainAreaHeight(pos1.GetX(), pos0.GetY(), exists[1]),
    GetTerrainAreaHeight(pos0.GetX(), pos1.GetY(), exists[2]),
    GetTerrainAreaHeight(pos1.GetX(), pos1.GetY(), exists[3]),
};
```

**Review by invertednormal — APPROVED** [review]
Looks good, just a few suggestions.

File: Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp [hunk 4]
```diff
@@ -223,6 +223,72 @@ void TerrainSystem::RoundPosition(float x, float y, float queryResolution, AZ::V
...
+    {
+        if (lerpY < 0.5f)
+        {
+            outExists = exists[2];
+        }
+        else
+        {
+            outExists = exists[3];
+        }
+    }
```
[the remark this answers is not in the document]
Comment (mbalfour-amzn) [reply Gems/Terrain/Code/Source/TerrainSystem/TerrainSystem.cpp]: (Small correction - the lerpY and lerpX are reversed in the suggestion above)

**Review by cgalvan — APPROVED** [review]

**Pull request merged**